### PR TITLE
Allow TABPOT type pairwise potential

### PR DESCRIPTION
**Pull request by k9cdt**
Added feature to allow using a LAMMPS style input for pairwise potentials. In this file lists the potential and forces in a table, allowing maximum flexibility in some cases.

The first draft of the code was written by @alexmir18 when he was a postdoc in Voth Group at UChicago.

### CODE REVIEW

**Comment by oschuett** [comment]
Out of curiosity, why did you withdraw this PR?

At first glance it looks good to me. My only comments would be:
 - please run `make pretty`
 - please add at least one test
 - please use `tap` consistently (some names use `tp`)

On a related note, the [GENPOT](https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL/MM/FORCEFIELD/NONBONDED/GENPOT.html) also allows for (almost) arbitrary potentials.

**Comment by k9cdt** [comment]
Yes I am thinking about adding tests. So far everything is only theoretically implemented correct. I will do the make pretty as well.

File: src/pair_potential_types.F [hunk 1]
```diff
@@ -49,9 +49,10 @@ MODULE pair_potential_types
                                  siepmann_type = 13, &
                                  gal_type = 14, &
                                  quip_type = 15, &
-                                 gal21_type = 16
+                                 gal21_type = 16, &
+                                 tp_type = 17
```
Comment (oschuett) [inline src/pair_potential_types.F]: Please rename `tp_type`  -> `tab_type`.

File: src/pair_potential_types.F [hunk 2]
```diff
@@ -300,6 +304,15 @@ MODULE pair_potential_types
       LOGICAL        :: express
    END TYPE gal21_pot_type
 
+! **************************************************************************************************
+
+   TYPE tp_pot_type
```
Comment (oschuett) [inline src/pair_potential_types.F]: Please rename `tp_pot_type` -> `tab_pot_type`.

File: src/pair_potential_types.F [hunk 3]
```diff
@@ -319,6 +332,7 @@ MODULE pair_potential_types
       TYPE(gal_pot_type), POINTER :: gal
       TYPE(gal21_pot_type), POINTER :: gal21
       TYPE(ftd_pot_type), POINTER :: ftd
+      TYPE(tp_pot_type), POINTER :: tp
```
Comment (oschuett) [inline src/pair_potential_types.F]: Please rename field: `tp` -> `tab`.

File: src/pair_potential_types.F [hunk 4]
```diff
@@ -1000,21 +1021,22 @@ END SUBROUTINE pair_potential_p_copy
 !> \param siepmann ...
 !> \param gal ...
 !> \param gal21 ...
+!> \param tp ...
 !> \author Teodoro Laino [Teo] 11.2005
 ! **************************************************************************************************
    SUBROUTINE pair_potential_reallocate(p, lb1_new, ub1_new, lj, lj_charmm, williams, &
                                         goodwin, eam, quip, bmhft, bmhftd, ipbv, buck4r, buckmo, gp, tersoff, &
-                                        siepmann, gal, gal21)
+                                        siepmann, gal, gal21, tp)
```
Comment (oschuett) [inline src/pair_potential_types.F]: Please rename arguments and variables `tp` -> `tab` everywhere.

**Comment by oschuett** [comment]
It seems your git history got pretty banged up (don't worry, it happens to all of us ;-).

The simplest solution is probably to **first backup your files**, then run:

```
git fetch origin
git reset --hard origin/master
```

... and then copy your modified files back in.

Alternatively, you can have a look at https://github.com/cp2k/cp2k/wiki/Git-Tips-&-Tricks

**Comment by oschuett** [comment]
Thanks a lot for contributing!

**Comment by alexmir18** [comment]
I am a developer of a TABPOT feature in CP2K (Voth group, University of Chicago, 2018-2020). Thanks to [k9cdt](https://github.com/k9cdt) for adding tests, making it clean, and merging it with the main branch.

**Pull request merged**